shadowsocks/udp.go

```diff
@@ -186,46 +186,79 @@ func (s *udpService) Stop() error {
 	return s.clientConn.Close()
 }
 
-type natentry struct {
-	conn net.PacketConn
+const (
+	modeFresh     = 0    // No packets have been sent.
+	modeSingleDNS = iota // One packet has been sent, a DNS query.
+	modeMultiDNS  = iota // Multiple packets have been sent, all DNS.
+	modeGeneral   = iota // Non-DNS packets have been sent.
+)
+
+type natconn struct {
+	net.PacketConn
 	// We store the client location in the NAT map to avoid recomputing it
 	// for every outbound packet in a UDP-based connection.
 	clientLocation string
+	// NAT timeout to apply for non-DNS connections.
+	defaultTimeout time.Duration
+	// Indicates whether DNS optimizations apply.
+	mode int32
+}
+
+func (c *natconn) WriteTo(buf []byte, dst net.Addr) (int, error) {
+	_, port, _ := net.SplitHostPort(dst.String())
+
+	// Update mode.
+	mode := c.mode
+	if port == "53" {
+		if mode == modeFresh {
+			mode = modeSingleDNS
+		} else if mode == modeSingleDNS {
+			mode = modeMultiDNS
+		}
+	} else {
+		mode = modeGeneral
+	}
+	atomic.StoreInt32(&c.mode, mode)
+
+	// Update deadline.
+	timeout := c.defaultTimeout
+	if mode != modeGeneral {
```

Comment:
We have two optimizations:
- Shorter deadline for DNS
- Quick close on response for DNS

Can we better decouple them?

The shorter deadline optimization can be accomplished by looking only at the port for each packet. It's stateless, so it's easier to understand and reason about (though you need to save the current deadline)

Then you can leave the mode for the quick close optimization only. This will make both optimizations independent of each other and easier to understand. If you tweak one, you know you won't be breaking the other.

Reply:
I was able to get rid of the mode and simplify the logic, using `sync.Once`.  This is also very efficient (lock-free).